tests/test_users.py

```diff
@@ -189,6 +189,188 @@ def test_form_creation(self):
+        self.client.get('/logout')
+        r = self.client.post('/forms/toggle',
+                            data={'hashid': form_endpoint},
+                            follow_redirects=True)
+        self.assertEqual(200, r.status_code)
+        self.assertTrue(Form.query.first().disabled)
+
+        # fail when attempting to post to form
+        r = self.client.post('/' + form_endpoint,
+            headers={'Referer': 'formspree.io'},
+            data={'name': 'bruce'}
+        )
+        self.assertEqual(403, r.status_code)
+        self.assertEqual(0, Form.query.first().counter)
+
+        # log back in and re-enable form
+        r = self.client.post('/login',
+            data={'email': 'hello@world.com',
+                  'password': 'friend'}
+        )
+        r = self.client.post('/forms/toggle',
+                            data={'hashid': form_endpoint},
+                            follow_redirects=True)
+        self.assertEqual(200, r.status_code)
+        self.assertFalse(Form.query.first().disabled)
+
+        # successfully post to form
+        r = self.client.post('/' + form_endpoint,
+            headers={'Referer': 'formspree.io'},
+            data={'name': 'bruce'}
+        )
+        self.assertEqual(1, Form.query.first().counter)
+
+    def test_form_and_submission_deletion(self):
+        # create and login a user
+        r = self.client.post('/register',
+            data={'email': 'hello@world.com',
+                  'password': 'friend'}
+        )
+        self.assertEqual(r.status_code, 302)
+        self.assertEqual(1, User.query.count())
+
+        # upgrade user
+        user = User.query.filter_by(email='hello@world.com').first()
+        user.upgraded = True
+        DB.session.add(user)
+        DB.session.commit()
+
+        # successfully create form
+        r = self.client.post('/forms',
+            headers={'Accept': 'application/json', 'Content-type': 'application/json'},
+            data=json.dumps({'email': 'hope@springs.com'})
+        )
+        resp = json.loads(r.data)
+        self.assertEqual(r.status_code, 200)
+        self.assertIn('submission_url', resp)
+        self.assertIn('hashid', resp)
+        form_endpoint = resp['hashid']
+        self.assertIn(resp['hashid'], resp['submission_url'])
+        self.assertEqual(1, Form.query.count())
+        self.assertEqual(Form.query.first().id, Form.get_with_hashid(resp['hashid']).id)
+
+        # post to form
+        r = self.client.post('/' + form_endpoint,
+            headers={'Referer': 'formspree.io'},
+            data={'name': 'bruce'}
+        )
+
+        # confirm form
+        form = Form.query.first()
+        self.client.get('/confirm/%s:%s' % (HASH(form.email, str(form.id)), form.hashid))
+        self.assertTrue(Form.query.first().confirmed)
+        self.assertEqual(0, Submission.query.count())
+
+        # increase the submission limit
+        settings.ARCHIVED_SUBMISSIONS_LIMIT = 10
```

Comment:
Oh yes this is true. When I originally looked at the setup code I guess I thought it would tear everything down again and set it up, but now I realize that it's only setting up the database each time, not the submission limits and such, so I should change that because that's going to be very hard to debug otherwise.